DREAM/Optimization/tsgParticleSwarm.hpp

```diff
@@ -0,0 +1,92 @@
+ *    or promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * UT-BATTELLE, LLC AND THE UNITED STATES GOVERNMENT MAKE NO REPRESENTATIONS AND DISCLAIM ALL WARRANTIES, BOTH EXPRESSED AND
+ * IMPLIED. THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_PARTICLE_SWARM_SOLVER_HPP
+#define __TASMANIAN_PARTICLE_SWARM_SOLVER_HPP
+
+#include "tsgOptimizationUtils.hpp"
+
+namespace TasOptimization {
+
+class ParticleSwarmState {
+  public:
+    ParticleSwarmState() = delete;
+    ParticleSwarmState(int nd) : num_dimensions(nd), num_particles(0) {};
+
+    inline std::vector<double> getParticlePositions() const {return particle_positions;}
+    inline std::vector<double> getParticleVelocities() const {return particle_velocities;}
+    inline std::vector<double> getBestParticlePositions() const {return best_particle_positions;}
+    inline std::vector<double> getBestPosition() const {return best_position;}
+    inline int getNumIterations() const {return num_iterations;}
+    inline int getNumDimensions() const {return num_dimensions;}
+    inline int getNumParticles() const {return num_particles;}
+
+    void setParticlePositions(const std::vector<double> &pp);
+    void setParticlePositions(const std::vector<double> &&pp) {setParticlePositions(pp);};
+    void setParticleVelocities(const std::vector<double> &pv);
+    void setParticleVelocities(const std::vector<double> &&pv) {setParticleVelocities(pv);};
+    void setBestParticlePositions(const std::vector<double> &bpp);
+    void setBestParticlePositions(const std::vector<double> &&bpp) {setBestParticlePositions(bpp);}
```

Comment:
* rvalue reference should not be const here
* this still copies on the call to `setBestParticlePositions();`, the idea of the rvalue is to move without copy, you need a second version of `setBestParticlePositions();` that uses `best_position = std::move(bp);`

Reply:
Fixed. The checks in these set functions now come from the helper function `checkVarSize()` in `tsgOptimizationUtils.hpp`.